Don't "optimize" empty list to no list, as the latter is used to indicate
no "in [word ...]" part and causes getting the positional args.

Idea from OpenBSD list:
http://openbsd.7691.n7.nabble.com/pdksh-bug-on-an-empty-for-loop-preceded-by-a-set-td229391.html

--- pdksh-5.2.14/syn.c.orig	1999-04-16 20:28:49.000000000 +0200
+++ pdksh-5.2.14/syn.c	2014-08-21 16:12:54.026320226 +0200
@@ -606,13 +606,8 @@
 		XPput(args, yylval.cp);
 	if (c != '\n' && c != ';')
 		syntaxerr((char *) 0);
-	if (XPsize(args) == 0) {
-		XPfree(args);
-		return NULL;
-	} else {
-		XPput(args, NULL);
-		return (char **) XPclose(args);
-	}
+	XPput(args, NULL);
+	return (char **) XPclose(args);
 }
 
 /*
